src/Collateral.sol

```diff
@@ -0,0 +1,202 @@
+     * @notice The collateral must be approved for transfer by the sender.
+     * @notice REVERT: this function will revert if the collateral transfer fails.
+     */
+    function depositCollateral(address receiver, uint256 amount) external {
+        collateralAccounts[msg.sender][receiver].balance += amount;
+        emit Deposit(msg.sender, receiver, amount);
+        require(collateralToken.transferFrom(msg.sender, address(this), amount));
+    }
+
+    /**
+     * @dev Requests to thaw a specific amount of collateral from a receivers collateral account.
+     * @param receiver Address of the receiver the collateral account is for.
+     * @param amount Amount of collateral to thaw.
+     * @notice REVERT: this function will revert if the sender receiver collateral account does
+     *                 not have enough collateral (greater than `amount`).
+     */
+    function thawCollateral(address receiver, uint256 amount) external {
+        CollateralAccount storage account = collateralAccounts[msg.sender][receiver];
+        uint256 totalThawingRequested = account.amountThawing + amount;
+        require(account.balance >= totalThawingRequested, "Insufficient collateral balance");
+
+        // Increase the amount being thawed
+        account.amountThawing = totalThawingRequested;
+        // Set when the thaw is complete (thawing period number of blocks after current block)
+        account.thawEnd = block.number + thawingPeriod;
+
+        emit ThawRequest(msg.sender, receiver, amount, account.thawEnd);
+    }
+
+    /**
+     * @dev Withdraws all thawed collateral from a receivers collateral account.
+     * @param receiver Address of the receiver.
+     * @notice REVERT: this function will revert if the sender receiver collateral account does
+     *                not have any thawed collateral. This function will also revert if no thawing
+     *               period has been completed.
+     */
+    function withdrawThawedCollateral(address receiver) external {
+        CollateralAccount storage account = collateralAccounts[msg.sender][receiver];
+        require(account.thawEnd != 0, "No collateral thawing");
+        require(account.thawEnd <= block.number, "Collateral still thawing");
+
+        // Amount is the minimum between the amount being thawed and the actual balance
+        uint256 amount = account.amountThawing > account.balance ? account.balance : account.amountThawing;
+
+        unchecked {
+            account.balance -= amount; // Reduce the balance by the withdrawn amount (no underflow risk)
+        }
+        account.amountThawing = 0;
+        account.thawEnd = 0;
+        emit Withdraw(msg.sender, receiver, amount);
+        require(collateralToken.transfer(msg.sender, amount));
+    }
+
+    /**
+     * @dev Authorizes a signer to sign RAVs for the sender.
+     * @param signer Address of the authorized signer.
+     */
+    function authorizeSigner(address signer) external {
```

Comment:
`authorizeSigner` being permissionless allows for a griefing attack where an attacker could block signer addresses by front-running calls and setting their own address as sender. 


Two potential solutions to this:
 1. Require proof of signer ownership before setting `authorizedSigners[signer]`, with a message signed by the signer, similar to what we suggested for the allocation id’s.
 2. Change the mapping to include the sender `authorizedSigners[sender][signer]`, however this would also require some way to retrieve the `sender` on the `Collateral.redeem()` call which is probably not practical.

Reply:
Good catch. I think option 1 is probably the best option here.